Removing ndsl.Namelist from PyFV3

jjuyeonkim opened the pull request:
**Description**
This follows from [NDSL PR #246](https://github.com/NOAA-GFDL/NDSL/pull/246) and is part of [NDSL Issue # 64](https://github.com/NOAA-GFDL/NDSL/issues/64).

This PR attempts to remove `ndsl.Namelist` from PyFV3 by doing the following:
- Modifies`DynamicalCoreConfig.from_f90nml()` to use only `f90nml.Namelist` via NDSL helper functions
- Removes `DynamicalCoreConfig.from_namelist()`
- Modifies the translate tests to shed its reliance on `ndsl.Namelist` in favor the `DynamicalCoreConfig`
- Adds `--no_legacy_namelist` to the github workflow `translate.yaml`

Partial work on [NDSL Issue # 64](https://github.com/NOAA-GFDL/NDSL/issues/64)

**How Has This Been Tested?**
- Existing unit tests
- One new simple unit test that checks out creating the DynamicalCoreConfig from a nml file with a "namelist_override" 

**Checklist:**
- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules - There is a related Pace PR that needs to be checked in first for the pace tests to pass: https://github.com/NOAA-GFDL/pace/pull/152
- [x] New check tests, if applicable, are included
- [ ] Targeted model if this changed was triggered by a model need/shortcoming

jjuyeonkim committed:
Initial changes: Namelist replaced by dict

jjuyeonkim committed:
Creating utility function dycore_config_from_f90nml for creating DynamicalCoreConfig from f90nml.Namelist; using it in translate tests

jjuyeonkim committed:
Instead of ndsl.Namelist, using f90nml.Namelist + dycore_config_from_f90nml util function

jjuyeonkim committed:
Removing from_namelist and from_f90nml from DynamicalCoreConfig; Adding utility function to create DynamicalCoreConfig from f90nml.Namelist using default groups or all

jjuyeonkim committed:
GEOS wrapper work around

jjuyeonkim committed:
Using self.config rather than self.namelist in TranslateDynCore

jjuyeonkim committed:
Linting and small tweaks

jjuyeonkim committed:
Adding --f90nml_namelist_only flag

jjuyeonkim committed:
more linting + comment revisions

jjuyeonkim committed:
Changed translate test flag: --legacy_namelist_support=False

jjuyeonkim committed:
Using --no_legacy_namelist flag for translate tests

jjuyeonkim committed:
Adding target_groups to be parameter for dycore_config_from_f90nml

jjuyeonkim committed:
Cleanup: mainly using f90nml.Namelist where appropriate,
Using dycore_config_from_f90nml

jjuyeonkim committed:
Simple unit test for namelist_override + comment tweaks

jjuyeonkim committed:
Merge branch 'develop' into 20250925_namelist_helper

romanc reviewed (commented):
Thanks, Janice. That looks like a step in the right direction to me. I agree that `DynamicalCoreConfig.from_namelist()` should go away. And I think it's smart to store the dycore config in test classes as `self.config`.

I'm curious to hear why you decided to have `dycore_config_from_f90nml` as a standalone function instead of updating `DynamicalCoreConfig.from_f90nml()`. It seems to me that what ever is done in `dycore_config_from_f90nml()` could just move into `DynamicalCoreConfig.from_f90nml()`, no? Maybe I'm missing something.

jjuyeonkim committed:
Adding from_f90nml back to DynamicalCoreConfig
Removing utils/namelist.py
Using DynamicalCoreConfig.from_f90nml in translate tests
Tweaking override unit test

jjuyeonkim commented on tests/main/test_config_from_f90nml.py:
```diff
@@ -0,0 +1,64 @@
+    return [main_file_path, override_file_path]
+
+
+def test_config_from_f90nml_with_override():
```
This is a really simple test for trying out the override_namelist. I didn't see one that tested it out, so I didn't think it would hurt.

However: 
1. I don't know if it goes here or in the Pace unit tests. It's not currently being run during the CI because there's another test in the same directory that also isn't being run and may be outdated.
2. I'm hoping I got the functionality of the test right -- that the namelist_override completely overrides the previous namelist?

If no one knows, I'll just remove the test.

Edit: Ultimately, i just removed the unit test for now.

jjuyeonkim committed:
DynamicalCoreConfig from_dict, from_f90nml mods

jjuyeonkim committed:
Merge branch 'develop' into 20250925_namelist_helper

jjuyeonkim committed:
Removing simple unit test

romanc commented on tests/savepoint/translate/translate_fvsubgridz.py:
```diff
@@ -176,17 +178,16 @@ def __init__(
         for qvar in utils.tracer_variables:
             self.ignore_near_zero_errors[qvar] = True
         self.stencil_factory = stencil_factory
-        self.namelist = namelist
```
I think this should be replaced with
```py
self.config = DynamicalCoreConfig.from_f90nml(namelist)
```
because `class TranslateFVSubgridZ(ParallelTranslateBaseSlicing)` and `ParallelTranslateBaseSlicing` doesn't inherit from `TranslateDycoreFortranData2Py` where you added the above line in the `__init__`, so I think `self.config` (as used below) will be undefined and if people actually try to run this test it would thus fail

jjuyeonkim replied:
I think you're right. I added the `self.config` line.

romanc commented on tests/savepoint/translate/translate_cubedtolatlon.py:
```diff
@@ -31,7 +33,7 @@ def __init__(
         }
         self.stencil_factory = stencil_factory
-        self.grid_type = namelist.grid_type
+        self.grid_type = self.config.grid_type
```
I don't think this is going to work unless you add
```py
self.config = DynamicalCoreConfig.from_f90nml(namelist)
```
above because `class TranslateCubedToLatLon(ParallelTranslate2Py)` doesn't inherit from `TranslateDycoreFortranData2Py` where this is done automatically. I thus expect the test to fail if people were actually running it ...

jjuyeonkim replied:
I think you're right. I'll update this. I still can't get the test to run properly (it's getting skipped, but I haven't figured out why yet), but I shouldn't make it worse.

romanc commented on pyfv3/_config.py:
```diff
@@ -286,107 +293,59 @@ def __post_init__(self):
+    def from_dict(
+        cls,
+        data: dict,
+    ) -> "DynamicalCoreConfig":
```
```suggestion
    ) -> DynamicalCoreConfig:
```

You can add `from __future__ import annotations` at the top of the file and then you don't need to add quotes around types any more that the old type resolver couldn't resolve

jjuyeonkim replied:
I'll update this.

Edit -- NOTE: I made a slight [tweak](https://github.com/NOAA-GFDL/pace/pull/152) to Pace's `tests/main/fv3core/test_config.py::test_types_match` test to account for this change.

romanc commented on pyfv3/_config.py:
```diff
@@ -286,107 +293,59 @@ def __post_init__(self):
+        data: dict,
+    ) -> "DynamicalCoreConfig":
+        """Uses the nml to create a DynamicalCoreConfig, using the data
+        dictionary.
```
There's no nml anymore at this stage. Maybe something like
```suggestion
        """Create a DynamicalCoreConfig from the given data.
```
is more appropriate?

romanc commented on pyfv3/_config.py:
```diff
@@ -286,107 +293,59 @@ def __post_init__(self):
+    ) -> "DynamicalCoreConfig":
+        """Uses the nml to create a DynamicalCoreConfig.
+        Only the DEFAULT_DYCORE_NML_GROUPS from the nml are considered
+        when initializing the DynamicalCoreConfig.
```
This has changed, right? You can specify `target_groups` and it will consider those, no?

jjuyeonkim replied:
You're right. I'll fix this.

romanc commented on pyfv3/_config.py:
```diff
@@ -286,107 +293,59 @@ def __post_init__(self):
                 print(f"{self.namelist_override} does not exist")
                 raise
-            dycore_config = self.from_f90nml(f90_nml)
+            dycore_config = self.from_f90nml(f90_nml, self.target_nml_groups)
```
passing `self.*` as an argument to a class function of the same class is always a bit fishy. Because it's in the same class, the function actually already has access to `self.*`, so if you have to pass it as an argument, generally speaking, there's room for improvement in the design. And as far as I understand, this is temporary. So I'm fine with it.

jjuyeonkim replied:
Yeah, I noticed this as well. I think this is one of the things I left as is because I wasn't sure about how the namelist_override was supposed to be used, and I don't have many people to ask right now. At this point, I want to wait on changing this part. (Unfortunately, I don't know how temporary this is.) I'll at least add a comment in the code that this should probably be changed eventually.

romanc replied:
yeah, it's all good for now.

romanc reviewed (changes requested):
Thanks, Janice! I think we are moving the right direction. A couple of smaller things inline.

I think some tests will break because the don't automatically get `self.config` added. The class hierarchy for translate tests is a mess (and I'm not sure we get to clean it up anytime soon).

And while we're at it, I'd suggest to take `TranslateDycoreFortranData2Py` out of `translate_fvdynamics.py` because the class isn't used there. I'd suggest to put it in a separate file in the same folder and then adjust the path in the `__init__.py` such that imports like "from pyfv3.testing import TranslateDycoreFortranData2Py" don't need any change.

jjuyeonkim committed:
Adding self.config to TranslateCubedToLatLon

jjuyeonkim committed:
Update pyfv3/_config.py

comment fix

Co-authored-by: Roman Cattaneo <romanc@users.noreply.github.com>

jjuyeonkim committed:
Comment fix

jjuyeonkim committed:
Merge branch '20250925_namelist_helper' of github.com:jjuyeonkim/PyFV3 into 20250925_namelist_helper

jjuyeonkim committed:
Using 'from __future__ import annotations'

jjuyeonkim committed:
Adding note about fishy behavior in dycore config post init

jjuyeonkim committed:
Adding self.config into translate tests; Moving TranslateDycoreFortranData2Py into its own translate_data.py

romanc reviewed (approved):
Thanks, Janice. From my point of view, this is good to go :rocket:
